File: OS/PyPileAndFile/Pile.py
import Noeud as Noeud


class Pile:
    def __init__(self):
        self.deb = None

    def empiler(self, val):
        new = Noeud.Noeud(val)
        new.next = self.deb
        self.deb = new

    def depiler(self):
        if self.estVide():
            return None
        toRemove = self.deb
        self.deb = toRemove.next
        val = toRemove.val
        del toRemove
        return val

    def estVide(self):
        return self.deb == None

    def __repr__(self):
        temp = Pile()
        representer = "DEB--->"
        while self.deb:
            temp.empiler(self.depiler())
            representer += str(temp.deb.val) + "--->"
        representer += "END"
        while temp.deb:
            self.empiler(temp.depiler())
        return representer

    def depilerKElet(self, k):
        i = 0
        while i < k and not (self.estVide()):
            self.depiler()
            i += 1

    def depilerElt(self, elt):
        while not (self.estVide()):
            if self.depiler() == elt:
                break

    def contient(self, elet):
        if self.estVide():
            return False
        temp = Pile()
        contient = False
        while not (self.estVide()):
            temp.empiler(self.depiler())
            if elet == temp.deb.val:
                contient = True
                break
        while not (temp.estVide()):
            self.empiler(temp.depiler())
        return contient

File: OS/CPileAndFile/src/pile.c
#include<stdio.h>
#include<stdlib.h>
#include"../Headers/pile.h"


void testPile(void){
	pile*p= initialisePile();
	printf("Le pile est-t-il vide =%d [Y/N]=[1,0]:",isEmpty(p));
	printf("\nEmpiler\n");
	for (int i=0;i<5;i++)empiler(p,i);	
	printf("Le pile est-t-il vide =%d [Y/N]=[1,0]:",isEmpty(p));
	Afficher(p);
	printf("Depiler 3 elements");
	depilerKElt(p,3);
	Afficher(p);
	printf("depiler jusqu'a 0 \n");
	depilerElt(p,0);
	Afficher(p);
	}

noeud* creatNoeud(int val){
	noeud* new=(noeud*) malloc(sizeof(noeud));
	new->val=val;
	new->next=NULL;
	return new;
	}
	
	
	
pile* empiler(pile* p,int val){
	noeud* new=creatNoeud(val);
	new->next=p->deb;
	p->deb=new;
	return p;
}

int isEmpty(pile* p){
	 //1 si vide 0 si non
	 int isVide=0;
	 if (p->deb==NULL)isVide=1;
	 return isVide;
 }
 

int depiler(pile* p){
	if (isEmpty(p)){
	 return -1; }
	int val= p->deb->val;
	noeud* toRemove=p->deb;
	p->deb=p->deb->next;
	free(toRemove);
	return val;
	}

pile* initialisePile(){
	// Initialisation d'un pile
	pile* p;
	p=(pile*) malloc(sizeof(pile));
	p->deb=NULL;
	return p;
}
int Afficher(pile* p){
	if(isEmpty(p))return 0;
	pile* temp=initialisePile();
	printf("\nDEB ---> ");
	while (!isEmpty(p)){
	int val;
	val= depiler(p);
	empiler(temp,val);
	printf("%d ---> ",val);
	}
	while (!isEmpty(temp))	empiler(p,depiler(temp));
	printf("NULL\n");
	return 1;
	}
void depilerKElt(pile* p,int k){
	int i=0;
	while(!isEmpty(p)&& i<k){
	depiler(p);
	i++;
	};
}
void depilerElt(pile* p, int elt){
	while(!isEmpty(p)&& elt!=depiler(p));
}

